src/vs/base/browser/ui/dialog/dialog.ts

```diff
@@ -477,22 +477,45 @@ export class Dialog extends Disposable {
 	}
 
 	private rearrangeButtons(buttons: Array<string>, cancelId: number | undefined): ButtonMapEntry[] {
-		const buttonMap: ButtonMapEntry[] = [];
-		if (buttons.length === 0) {
-			return buttonMap;
-		}
 
-		// Maps each button to its current label and old index so that when we move them around it's not a problem
-		buttons.forEach((button, index) => {
-			buttonMap.push({ label: button, index });
-		});
+		// Maps each button to its current label and old index
+		// so that when we move them around it's not a problem
+		const buttonMap: ButtonMapEntry[] = buttons.map((label, index) => ({ label, index }));
+
+		if (buttons.length < 2) {
+			return buttonMap; // only need to rearrange if there are 2+ buttons
+		}
 
-		// macOS/linux: reverse button order if `cancelId` is defined
 		if (isMacintosh || isLinux) {
-			if (cancelId !== undefined && cancelId < buttons.length) {
+
+			// Linux: the GNOME HIG (https://developer.gnome.org/hig/patterns/feedback/dialogs.html?highlight=dialog)
+			// recommend the following:
+			// "Always ensure that the cancel button appears first, before the affirmative button. In left-to-right
```

Comment:
So we are not conforming with the Linux HIG then? As I read the code, you take the cancel button from wherever it is (line 504), and put it in the second position (line 505) and then reverse (line 508). This would put it in the second last position, which seems to be common in native OS X apps, but for > 2 buttons does not confirm with Linux HIG?

Reply:
The intent is to do:
* `[Cancel] [Default]`
* `[Foo] [Cancel] [Default]`
* `[Foo] [Bar] [Cancel] [Default]`

In other words, `Cancel` always being the second last one before the default button.

Reply:
Okay, I guess one can read he Linux HIG in multiple ways. They only have an example with two buttons. I read "Rule #1: Cancel should always be first PERIOD", but together with the second part of the sentence, I guess one can also read it as "Cancel should come before the affirmative button, and any tertiary buttons are not mentioned and thus undefined". They do mention that you can have up to three buttons but do not specify where the third one should go? Bummer.

Reply:
![image](https://user-images.githubusercontent.com/29040/216100663-29aee7b8-b19c-4c6e-808d-7999529741cd.png)

I tried the GEdit Text editor on my workstation, and they indeed have the cancel button in the middle. Still feels a bit odd to me, but I guess I let the case rest :)

Reply:
It is probably also worth nothing that we have been running with this for 5+ years with probably millions of Linux users and so far I have gotten 0 people complaining. Same is true for macOS btw.

Reply:
During testing I see that when we have 4 buttons, we do not properly sort the "Cancel" button.

Reply:
![image](https://user-images.githubusercontent.com/900690/216115405-5106bf6e-3113-4a2d-8a30-8ba1be1e67b9.png)